### PR TITLE
avm2: Initial implementation of URLStream

### PR DESCRIPTION
**Pull request by Aaron1011**
This is based on URLLoader, and doesn't actualy 'stream' data - it all becomes available once the request finishes. However, this is good enough to get Sniper Team working.

### CODE REVIEW

File: core/src/avm2/globals/flash/net/URLStream.as [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        public function get bytesAvailable():uint {
+            return this._loader.data.bytesAvailable;
+        }
+
+        public function get endian():String {
+            return _endian;
+        }
+
+        public function set endian(value:String):void {
+            _endian = value;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/net/URLStream.as:49]: See https://github.com/ruffle-rs/ruffle/pull/10879 for the validation:
```
        public function set endian(value:String):void {
            if (value === Endian.BIG_ENDIAN || value === Endian.LITTLE_ENDIAN) {
                this._endian = value;
                if (this._loader.data) {
                    this._loader.data.endian = value;
                }
            } else {
                throw new ArgumentError("Error #2008: Parameter endian must be one of the accepted values.", 2008);
            }
        }
```

File: core/src/avm2/globals/flash/net/URLStream.as [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    import flash.events.SecurityErrorEvent;
+    import flash.events.ProgressEvent;
+    import flash.net.URLLoaderDataFormat;
+    import flash.events;
+    import __ruffle__.stub_method;
+    import __ruffle__.stub_getter;
+    import __ruffle__.stub_setter;
+
+    public class URLStream extends EventDispatcher implements IDataInput {
+        private var _endian:String = Endian.LITTLE_ENDIAN;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/net/URLStream.as:17]: The docs say the default endian should be `BIG_ENDIAN`.

**Comment by Lord-McSweeney** [comment]
The `adobe.ac.pmd.rules.event.ConstructorDispatchingEvent` needs to be disabled here since it's assuming that the `dispatchEvent` logic is running inside the constructor, while it is instead running inside an event listener inside the constructor.

File: core/src/avm2/globals/flash/net/URLStream.as [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        private var _endian:String = Endian.BIG_ENDIAN;
+        private var _loader:URLLoader = new URLLoader();
+
+        public function URLStream() {
+            this._loader.dataFormat = URLLoaderDataFormat.BINARY;
+            var self = this;
+
+            this._loader.addEventListener(Event.COMPLETE, function(e:*):void {
+                self._loader.data.endian = self._endian;
+                self.dispatchEvent(new Event(Event.COMPLETE)); // NOPMD ConstructorDispatchingEvent
```
Comment (Dinnerbone) [inline core/src/avm2/globals/flash/net/URLStream.as:26]: With #11123 landing, we're removing all NOPMD comments now

File: core/src/avm2/globals/flash/net/URLStream.as [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    import __ruffle__.stub_method;
+    import __ruffle__.stub_getter;
+    import __ruffle__.stub_setter;
+
+    public class URLStream extends EventDispatcher implements IDataInput {
+        private var _endian:String = Endian.BIG_ENDIAN;
+        private var _loader:URLLoader = new URLLoader();
+
+        public function URLStream() {
+            this._loader.dataFormat = URLLoaderDataFormat.BINARY;
```
Comment (Dinnerbone) [inline core/src/avm2/globals/flash/net/URLStream.as:21]: Can you drop a comment in the constructor to explain that this is temporary implementation and not what we expect this to do in the future? Even just a `// HACK, see https://github.com/ruffle-rs/ruffle/pull/11046` would be sufficient

**Review by Dinnerbone — CHANGES_REQUESTED** [review]

File: core/src/avm2/globals/flash/net/URLStream.as (hunk 4 above)
Comment (Aaron1011) [reply core/src/avm2/globals/flash/net/URLStream.as:27]: I added a comment to the `URLLoader` field.

**Review by Dinnerbone — APPROVED** [review]

**Comment by Aaron1011** [comment]
@Dinnerbone I added a test, and made URLLoader fire a "progress" event.

**Pull request merged**